WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -9,46 +9,176 @@ import Alamofire
+        // Sign in to an existing account.
+        // This is typically used when the app needs to get a new OAuth token because the current one becomes invalid.
+        case reAuthentication(TaggedManagedObjectID<WPAccount>)
+        // Connect a site to Jetpack or sign in to an already connected site.
+        case jetpackSite(accountEmail: String?)
+
+        func accountEmail(in context: NSManagedObjectContext) -> String? {
+            switch self {
+            case .default:
+                return nil
+            case let .reAuthentication(accountID):
+                return (try? context.existingObject(with: accountID))?.email
+            case let .jetpackSite(email):
+                return email
+            }
+        }
+    }
+
+    enum AuthenticationError: Error {
         case invalidCallbackURL
         case loginDenied(message: String)
         case obtainAccessToken
         case urlError(URLError)
         case parsing(DecodingError)
         case cancelled
-        case unknown(Swift.Error)
+        case unknown(Error)
+    }
+
+    enum SignInError: Error {
+        case authentication(AuthenticationError)
+        case fetchUser(Error)
+        case mismatchedEmail(expectedEmail: String)
+        case alreadySignedIn(signedInAccountEmail: String)
+        case loadingSites(Error)
+    }
+
+    private let coreDataStack: CoreDataStackSwift
+    private let authenticator: ((URL) throws(AuthenticationError) -> URL)?
+
+    init(
+        coreDataStack: CoreDataStackSwift = ContextManager.shared,
+        authenticator: ((URL) throws(AuthenticationError) -> URL)? = nil
+    ) {
+        self.coreDataStack = coreDataStack
+        self.authenticator = authenticator
     }
 
+    /// Sign in WP.com account.
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
     @MainActor
-    func signIn(from viewController: UINavigationController) async {
+    func signIn(from viewController: UIViewController, context: SignInContext) async -> TaggedManagedObjectID<WPAccount>? {
+        WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "start"])
+        do {
+            let account = try await attemptSignIn(from: viewController, context: context)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "success"])
+            return account
+        } catch {
+            present(error, from: viewController)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "error", "error": "\(error)"])
+            return nil
+        }
+    }
+
+    /// Sign in WP.com account.
+    ///
+    /// - SeeAlso: `signIn`
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
+    @MainActor
+    func attemptSignIn(from viewController: UIViewController, context: SignInContext) async throws(SignInError) -> TaggedManagedObjectID<WPAccount> {
```

Comment:
(nit) should be private

Reply:
This function is called from unit tests.